### ISSUE TITLE
some reflections

### ISSUE THREAD

**Issue by cristianrep** [useaible/RyskampLearningMachine#17]
I got intrigued, by the algorithm and the samples and I decided to give it a try.
The samples worked wonderfully.

So I decided to give it a try for a text classification task. I had around 2k text docs  which were vectorized and ran with a 95% accuracy and F1Score = 85% using cntk neural network library with 80/20 training to testing.
I couldn't make it converge no matter what hyper parameters I used for linearity brackets.
I started debugging the code and realized that, in case of all vector features are linear, the algorithm is executing some sort of k neighbors variation. Now is not even looking for the closest one as distance or use  a successful solution for that neuron in the past in a random run. And like any k neighbor is supposed to compare all existing neurons before it takes any decision, which is very slow and also takes in consideration only one neuron which is a very rough estimation, there is no smoothing through considering more than one neighbor.  I guess I could have played more with the cycle score but at the price of more computation which would have made things even slower.
I guess the algorithm is a good solution for mixed distinct , linear combination features for a vector, but this results in feature engineering.

I'm writing to understand if I'm missing something.
Other than that the samples impressed me

All the best 
Cristian

**Comment by RixRyskamp** [comment]
Hi Cristian,

I am attaching what we call our differentiation vector.  You will see that
classification is not something we claim or plan to do well.  There are so
many engines that do this well already.  Our focus is on business
operational problems that cannot be solved by statistical engines.

Try the engine with a problem that requires finding the best solution with
a number of rules or non-linear non-statistical restrictions.  Stochastic
problems are great.  If you think about it, most business problems are not
as simple as classification in a purely academic sense.  You are required
to follow arbitrary human rules in most business problems.  This is what we
do well.

We often cite a project we did for Walmart where the best item to display
on a particular shelf is dictated more bureaucratic rules than by math.
Compared to other engines, we were the ONLY engine that could solve that
problem.

I would suggest you find a problem that is not typically associated with
pure machine learning (often something where people use rules engines after
and before the ML components) and try it.  These practical applications are
where we are designed to shine.

I am interested in your feedback and happy to answer more questions.  We
can direct you to the whitepaper and code for the above Walmart problem if
you did not find it on your own.  Thanks for playing with our product!

Thanks,

Rix Ryskamp

**Comment by cristianrep** [comment]
Thanks for answering 
Is it fair to call it a combination in between random forest and k nearest (best approximate name I could come with ) that allows both running under the same hood.

i'll have  a look at Walmart study case 

Thanks
Cristian

**Comment by RixRyskamp** [comment]
It is funny, we often get asked to compare the RLM to different
algorithms.  I built it from scratch with no ties to existing algorithms.
It is really its own entity.  It does share attributes with many other
machine learning systems, such as genetic, k nearest neighbor, neural nets,
forests, and others.  This is a coincidence more than a design feature.

However, the RLM was built to solve business problems that contain rules
that statistics alone are not ideal for solving while staying 100% machine
learning (and keeping the benefits of machine learning).  Another
interesting note is that our system functions much more effectively in
unsupervised learning. By default, it wants to learn as it goes, not be
pre-fed information.  It is a really its own beast.  It is great for
certain things (primarily business operations) and for others we refer
people to TensorFlow or other traditional solutions, as you learned with
classification.

Thanks,

Rix

**Comment by cristianrep** [comment]
Designing with your goal in mind is definitely the best.

I can probably make RLM  work even for my case, by introducing a distinct type feature (just used as a flag that could trigger the rule side of the algorithm) in the vector coming from a clustering previous step before getting consumed by RLM. It'll  have to compare for linearity only on the clustering subset then. The results would converge faster but the accuracy it'll be dictated somehow by the quality of the clustering.

Thanks for chatting with me,
Cristian

**Comment by RixRyskamp** [comment]
Me know if I can help in any way!
